osu.Game/Online/API/APIAccess.cs

```diff
@@ -208,11 +209,14 @@ private void attemptConnect()
 
             if (!authentication.HasValidAccessToken)
             {
+                state.Value = APIState.Connecting;
                 LastLoginError = null;
 
                 try
                 {
                     authentication.AuthenticateWithLogin(ProvidedUsername, password);
+                    state.Value = APIState.RequiresSecondFactorAuth;
```

Comment:
So this means _every_ user (who doesn't have their password enabled locally, I suppose) is going to have to 2FA verify on login? It's not like web wherein only certain user actions require verification?

Reply:
Yep, we're forcing it this time around at login. Seems more consistent.

Reply:
Yeah okay. I'm personally all for it because doing it as web does would be... rather painful...

That does beg the question though, what about users who have stored refresh tokens in config and as such will not go through the normal login flow? Is it presumed that the user is 2FA-verified throughout the entire lifetime of the refresh token? If so, are we going to have some possibility of revocation of the refresh token available to the user in event of account compromise?

Reply:
I've mentioned this in the OP:

> Response from a request (either error or otherwise) to let us know that verification hasn't been performed. ie. consider the case where the user does user/password login then quits the game before 2FA authenticating – to handle this, osu-web needs to let us know that we should show the flow again.

Basically there will be a case where web will let us know via a response to any API request that we need verification, which will cause the client to bring up the prompt.

This also means we could implement verify-when-required without any pain.